File: phone/acg-android/app/src/main/java/com/aicraftspeopleguild/acg/udt/UdtRegistry.kt
package com.aicraftspeopleguild.acg.udt

import android.content.Context
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive

/**
 * Reads every *.template.json + instance-*.json + *-*.template.json
 * sitting in app/src/main/assets/udt/ at boot, groups them by udtType,
 * and exposes a single queryable catalog.
 *
 * Screens can now iterate `registry.instances("AndroidModule")` to
 * draw "what's running" without hardcoding a Kotlin list.
 */
class UdtRegistry(ctx: Context) {
    private val json = Json { ignoreUnknownKeys = true }

    data class Template(val udtType: String, val file: String, val description: String)
    data class Instance(
        val udtType: String, val instance: String, val file: String,
        val parameters: JsonObject,
    ) {
        fun str(key: String): String? = parameters[key]?.jsonPrimitive?.content
        fun strList(key: String): List<String> =
            (parameters[key] as? JsonArray)?.map { it.jsonPrimitive.content } ?: emptyList()
    }

    val templates: List<Template>
    val instances: List<Instance>

    init {
        val am = ctx.assets
        val files = am.list("udt")?.toList().orEmpty()
        val tpls = mutableListOf<Template>()
        val insts = mutableListOf<Instance>()

        for (f in files) {
            val text = am.open("udt/$f").bufferedReader().use { it.readText() }
            val doc  = runCatching { json.parseToJsonElement(text).jsonObject }.getOrNull() ?: continue
            val udtType = doc["udtType"]?.jsonPrimitive?.content ?: continue

            if (f.endsWith(".template.json")) {
                tpls += Template(
                    udtType     = udtType,
                    file        = "udt/$f",
                    description = doc["description"]?.jsonPrimitive?.content ?: "",
                )
            } else {
                // Instances carry {udtType, instance, parameters}
                val inst = doc["instance"]?.jsonPrimitive?.content ?: continue
                val params = doc["parameters"] as? JsonObject ?: continue
                insts += Instance(udtType, inst, "udt/$f", params)
            }
        }
        templates = tpls.sortedBy { it.udtType }
        instances = insts.sortedWith(compareBy({ it.udtType }, { it.instance }))
    }

    fun templates(udtType: String): List<Template> = templates.filter { it.udtType == udtType }
    fun instances(udtType: String): List<Instance> = instances.filter { it.udtType == udtType }
    fun typesPresent(): List<String> = instances.map { it.udtType }.distinct().sorted()
    fun countByType(): Map<String, Int> =
        instances.groupingBy { it.udtType }.eachCount().toSortedMap()
}

File: phone/acg-ios/ACG/UDT/UdtRegistry.swift
import Foundation

/// Loads every JSON file in Resources/udt/ at startup. Files whose
/// names end in `.template.json` are UDT type declarations; every
/// other JSON is expected to carry `{udtType, instance, parameters}`
/// and lands in the instance list.
struct UdtRegistry {
    struct Template { let udtType: String; let file: String; let description: String }
    struct Instance {
        let udtType: String
        let instance: String
        let file: String
        let parameters: [String: Any]

        func str(_ key: String) -> String?     { parameters[key] as? String }
        func strList(_ key: String) -> [String] { (parameters[key] as? [String]) ?? [] }
    }

    let templates: [Template]
    let instances: [Instance]

    var typesPresent: [String] { Array(Set(instances.map { $0.udtType })).sorted() }
    var countByType: [String: Int] {
        Dictionary(grouping: instances, by: \.udtType).mapValues { $0.count }
    }

    func templates(of udtType: String) -> [Template] { templates.filter { $0.udtType == udtType } }
    func instances(of udtType: String) -> [Instance] {
        instances.filter { $0.udtType == udtType }.sorted { $0.instance < $1.instance }
    }

    static func load() -> UdtRegistry {
        guard let url = Bundle.main.url(forResource: "udt", withExtension: nil),
              let files = try? FileManager.default.contentsOfDirectory(at: url, includingPropertiesForKeys: nil)
        else { return UdtRegistry(templates: [], instances: []) }

        var tpls: [Template] = []; var insts: [Instance] = []
        for f in files where f.pathExtension == "json" {
            guard let data = try? Data(contentsOf: f),
                  let obj  = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
                  let udtType = obj["udtType"] as? String
            else { continue }

            if f.lastPathComponent.hasSuffix(".template.json") {
                tpls.append(.init(udtType: udtType, file: "udt/\(f.lastPathComponent)",
                                  description: obj["description"] as? String ?? ""))
            } else {
                guard let inst = obj["instance"] as? String,
                      let params = obj["parameters"] as? [String: Any] else { continue }
                insts.append(.init(udtType: udtType, instance: inst,
                                   file: "udt/\(f.lastPathComponent)", parameters: params))
            }
        }
        return UdtRegistry(
            templates: tpls.sorted { $0.udtType < $1.udtType },
            instances: insts.sorted { "\($0.udtType)\($0.instance)" < "\($1.udtType)\($1.instance)" },
        )
    }
}
